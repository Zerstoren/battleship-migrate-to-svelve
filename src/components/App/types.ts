import { ILobbyStore } from '../../stores/lobby';

export enum GameStatus {
  MAIN = 'main',
  LOBBY_LIST = 'list',
  CREATE_LOBBY = 'create',
  WAIT_CONNECT = 'wait',
  SET_SHIPS = 'set_ships',
  GAME = 'game_process',
  GAMEOVER = 'game_loose',
  GAMEWIN = 'game_win',
}

export enum FireTurn {
  NOBODY = 'nobody',
  ME = 'me',
  OPPONENT = 'opponent',
}

export interface IGameState {
  gameStatus: GameStatus,
  gameWithUser: boolean,

  setGameStatus: (status: GameStatus) => void,
}

export interface IProps {
  handleChangeGameStatus: (startGame: GameStatus, lobby?: ILobbyStore | null) => void,
}

export interface ISetLobbyProps {
  handleSetLobby: (startGame: GameStatus, lobby: ILobbyStore) => void
}
