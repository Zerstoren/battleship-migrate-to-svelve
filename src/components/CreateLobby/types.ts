export interface FormErrors {
  lobbyName?: string,
  x?: string,
  y?: string,
  shipsCount?: string,
  ships4n?: string,
  ships3n?: string,
  ships2n?: string,
  ships1n?: string,
}

export interface FormInputData {
  lobbyName: string,
  x: string,
  y: string,
  ships4n: string,
  ships3n: string,
  ships2n: string,
  ships1n: string,
}
