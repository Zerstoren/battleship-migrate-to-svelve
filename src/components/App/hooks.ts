import { useEffect } from 'react';
import client from '../../API';
import useWebsocketServer from '../../shared/hooks/websocketServer';
import { IMainStore } from '../../stores/mainStore';

const useWebsocketAppReconnect = (state: IMainStore) : void => {
  useWebsocketServer('opponentDisconnect', () => {
    state.setError('Disconnected');
  });

  useWebsocketServer('opponentReconnect', () => {
    state.setError('');
  });

  const reconnectAction = useWebsocketServer('reconnect', (data: {error?: string, success?: string}) => {
    if (data.error) {
      state.setError(data.error);
    } else {
      state.setError('');
    }
  });

  useEffect(() => {
    let lastId: string;

    const reconnectFn = () => {
      if (!lastId) {
        lastId = client.id;
        return;
      }

      reconnectAction({
        newId: client.id,
        oldId: lastId,
      });

      lastId = client.id;
    };

    const disconnectFn = () => {
      setTimeout(() => client.connect(), 1000);
      state.setError('Disconnected, try reconnect');
    };

    client.on('disconnect', disconnectFn);
    client.on('connect', reconnectFn);

    return () => {
      client.off('disconnect', disconnectFn);
      client.off('connect', reconnectFn);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
};

export default useWebsocketAppReconnect;
