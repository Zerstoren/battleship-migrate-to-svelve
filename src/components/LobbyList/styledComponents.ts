import styled from 'styled-components';

export const LobbyTable = styled.table.attrs(() => ({
  className: 'table table-striped table-hover',
}))``;
