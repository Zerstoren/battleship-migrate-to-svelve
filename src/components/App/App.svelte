<script lang="ts">
import { getContext } from 'svelte';
import type { IStore } from '../../stores';
import { STORE_CONTEXT } from '../../stores';
import CreateLobby from '../CreateLobby/CreateLobby.svelte';
import LobbyList from '../LobbyList/LobbyList.svelte';
import MainPage from '../MainPage/MainPage.svelte';
import SetShips from '../SetShips/SetShips.svelte';
import WaitConnect from '../WaitConnect/WaitConnect.svelte';
import { GameStatus } from './types';

const mainStore = getContext<IStore>(STORE_CONTEXT).mainStore;
let store = mainStore.store;

const handleAcceptError = () => {
  mainStore.setError('');
}

</script>

<div>
  {#if $store.error}
    <div class="alertError alert alert-warning alert-dismissible fade show">
      <strong>{$store.error}</strong>
      <input type="button" class="btn-close" onClick={handleAcceptError} />
    </div>
  {/if}
  <div class="centerDiv">
    {#if $store.status === GameStatus.MAIN}
      <MainPage />
    {:else if $store.status === GameStatus.CREATE_LOBBY}
      <CreateLobby />
    {:else if $store.status === GameStatus.WAIT_CONNECT}
      <WaitConnect />
    {:else if $store.status === GameStatus.LOBBY_LIST}
      <LobbyList />
    {:else if $store.status === GameStatus.SET_SHIPS}
      <SetShips />
    {/if}
  </div>
</div>

<style>
.centerDiv {
  width: 600px;
  margin: 0 auto;
  text-align: center;
}

.alertError {
  text-align: center;
}
</style>