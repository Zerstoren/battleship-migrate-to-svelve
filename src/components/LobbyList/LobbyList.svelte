<script lang="ts">
import { getContext, onMount } from "svelte";
import { STORE_CONTEXT } from "../../stores";
import useWebsocketServer from "../../shared/hooks/websocketServer";
import Header from "../../shared/StyledComponents/Header.svelte";
import { GameStatus } from "../App/types";
import { get} from "svelte/store";
import type { IStore } from "../../stores";
import type { ILobbySet, ILobbyData } from "../../stores/lobbyList";
import type { ILobby } from "../../stores/lobby";
import type { Readable }   from "svelte/store";

const {
  lobbyList,
  mainStore,
} = getContext<IStore>(STORE_CONTEXT);
const lobbyListStore = lobbyList.store;

const connectToLobby = useWebsocketServer('connectLobby');

const lobbyListRequest = useWebsocketServer('lobbyList', (data: ILobbyData) => {
  lobbyList.updateFromServer(data);
});

onMount(() => {
  lobbyListRequest({});
})

interface ILobbyArrayItem {
  userId: string,
  data: ILobby,
  lobby: Readable<ILobby>,
}

let lobbyListData: ILobbyArrayItem[] = [];

$: {
  lobbyListData = Object.entries($lobbyListStore).map((items) => {
    const [userId, lobby] = items;

    return {
      userId: userId,
      data: get(lobby),
      lobby: lobby,
    };
  });
}

const handleJoin = (user: string, lobby: Readable<ILobby>) => {
  connectToLobby(user);
  mainStore.setLobby(GameStatus.SET_SHIPS, lobby);
  console.log('Joined')
};

</script>

<Header>Lobby list</Header>

<table class="table table-striped table-hover">
  <thead>
    <tr>
      <td />
      <td>Lobby name</td>
      <td>Field size</td>
      <td>Ships 4x/3x/2x/1x</td>
    </tr>
  </thead>
  <tbody>
    {#each lobbyListData as lobbyItem}
      <tr>
        <td>
          <button class="btn btn-secondary" on:click={() => handleJoin(lobbyItem.userId, lobbyItem.lobby)}>Join</button>
        </td>
        <td>{lobbyItem.data.name}</td>
        <td>
          {lobbyItem.data.x} x {lobbyItem.data.y}
        </td>
        <td>
          {lobbyItem.data.ships4n} / {lobbyItem.data.ships3n} / {lobbyItem.data.ships2n} / {lobbyItem.data.ships1n}
        </td>
      </tr>
    {/each}
  </tbody>
</table>