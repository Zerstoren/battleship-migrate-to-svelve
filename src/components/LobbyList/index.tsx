import { inject, observer } from 'mobx-react';
import React, { FC, useEffect } from 'react';
import useWebsocketServer from '../../shared/hooks/websocketServer';
import { AppHeader } from '../../shared/StyledComponents/Headers';
import { IDataFromServer, ILobbyStore } from '../../stores/lobby';
import { ILobbyList } from '../../stores/lobbyList';
import { GameStatus, ISetLobbyProps } from '../App/types';
import { LobbyTable } from './styledComponents';

interface IProps extends ISetLobbyProps {
  lobbyList?: ILobbyList
}

const LobbyList: FC<IProps> = inject('lobbyList')(observer((props: IProps) => {
  const lobbyList = props.lobbyList as ILobbyList;
  const { handleSetLobby } = props;
  const connectToLobby = useWebsocketServer('connectLobby');

  const lobbyListRequest = useWebsocketServer('lobbyList', (data: IDataFromServer) => {
    lobbyList.updateFromServer(data);
  });

  useEffect(() => {
    lobbyListRequest({});
  }, [lobbyListRequest]);

  const handleJoin = (user: string, lobby: ILobbyStore) => {
    connectToLobby(user);
    handleSetLobby(
      GameStatus.SET_SHIPS,
      lobby,
    );
  };

  const items: JSX.Element[] = [];
  for (const record of lobbyList?.lobbys.values()) {
    items.push(
      <tr key={record.name}>
        <td>
          <button className="btn btn-secondary" onClick={() => handleJoin(record.name, record.value)}>Join</button>
        </td>
        <td>{record.value.lobbyName}</td>
        <td>
          {record.value.x}
          x
          {record.value.y}
        </td>
        <td>
          {record.value.ships4n}
          /
          {record.value.ships3n}
          /
          {record.value.ships2n}
          /
          {record.value.ships1n}
        </td>
      </tr>,
    );
  }

  return (
    <>
      <AppHeader>Lobby list</AppHeader>
      <LobbyTable>
        <thead>
          <tr>
            <td />
            <td>Lobby name</td>
            <td>Field size</td>
            <td>Ships 4x/3x/2x/1x</td>
          </tr>
        </thead>
        <tbody>
          {items}
        </tbody>

      </LobbyTable>
    </>
  );
}));

export default LobbyList;
