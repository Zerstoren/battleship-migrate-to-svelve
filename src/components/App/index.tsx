import { inject, observer } from 'mobx-react';
import React, { FC } from 'react';
import { ILobbyStore } from '../../stores/lobby';
import { IMainStore } from '../../stores/mainStore';
import CreateLobby from '../CreateLobby';
import EndGameScreen from '../EndGameScreen';
import Game from '../Game';
import LobbyList from '../LobbyList';
import Main from '../MainPage';
import SetShips from '../SetShips';
import WaitConnect from '../WaitConnect';
import useWebsocketAppReconnect from './hooks';
import { AlertError, CenteredDiv } from './styledComponents';
import { GameStatus } from './types';

interface IAppProps {
  mainStore?: IMainStore
}

const App: FC<IAppProps> = inject('mainStore')(observer((props) => {
  const state = props.mainStore as IMainStore;

  useWebsocketAppReconnect(state);

  const handleGameStart = (startGame: GameStatus) => state?.setGameStatus(startGame);
  const handleSetLobby = (startGame: GameStatus, lobby: ILobbyStore) => state?.setLobby(startGame, lobby);
  const handleWaitLobby = (
    startGame: GameStatus,
    lobby: ILobbyStore | null = null,
  ) => state.setLobby(startGame, lobby);

  const handleAcceptError = () => state?.setError('');

  let component = null;

  switch (state.status) {
    case GameStatus.MAIN:
      component = (<Main handleChangeGameStatus={handleGameStart} />);
      break;

    case GameStatus.CREATE_LOBBY:
      component = (<CreateLobby handleChangeGameStatus={handleWaitLobby} />);
      break;

    case GameStatus.WAIT_CONNECT:
      component = (<WaitConnect />);
      break;

    case GameStatus.LOBBY_LIST:
      component = (<LobbyList handleSetLobby={handleSetLobby} />);
      break;

    case GameStatus.SET_SHIPS:
      component = (<SetShips />); break;

    case GameStatus.GAME:
      component = (<Game />);
      break;

    case GameStatus.GAMEWIN:
    case GameStatus.GAMEOVER:
      component = (<EndGameScreen gameResult={state.status as GameStatus} />);
      break;

    default:
      break;
  }

  let err = null;

  if (state?.error) {
    err = (
      <AlertError>
        <strong>{state?.error}</strong>
        <input type="button" className="btn-close" onClick={handleAcceptError} />
      </AlertError>
    );
  }

  return (
    <>
      {err}
      <CenteredDiv>
        {component}
      </CenteredDiv>
    </>
  );
}));

export default App;
