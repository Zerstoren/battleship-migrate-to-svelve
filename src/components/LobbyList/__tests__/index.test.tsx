import React from 'react';
import {mount, ReactWrapper} from 'enzyme';
import { Provider } from 'mobx-react';

jest.mock('../../../shared/hooks/websocketServer', () => jest.fn());
import * as WebsocketServer from '../../../shared/hooks/websocketServer';

import { ILobbyList, LobbyList as LobbyListStore } from '../../../stores/lobbyList';
import LobbyList from '..';
import { applySnapshot } from 'mobx-state-tree';
import { ILobbyStore } from '../../../stores/lobby';
import { GameStatus } from '../../App/types';

const mockWebsocket = () => {
  const useWebsocket: jest.Mock = WebsocketServer.default as jest.Mock;
  
  type websocketFn = (data: Record<string, any>) => void;
  const websocketFns: Record<string, websocketFn> = {
    lobbyListFn: () => null,
    connectLobbyFn: () => null
  }

  const lobbyList = jest.fn();
  const connectLobby = jest.fn();
  useWebsocket.mockImplementation((path, fn) => {
    switch(path) {
      case 'lobbyList':
        websocketFns.lobbyListFn = fn;
        return lobbyList;
      
      case 'connectLobby':
        websocketFns.connectLobbyFn = fn;
        return connectLobby;

      default:
        throw new Error(`Path ${path} is not resolved`);
    }
  });

  return {
    lobbyList,
    connectLobby,
    websocketFns
  };
};

describe('LobbyList', () => {
  let tree: ReactWrapper;
  let fn: jest.Mock;
  let lobbyList: ILobbyList;

  beforeEach(() => {
    lobbyList = LobbyListStore.create();
    fn = jest.fn(() => null);
  });
  
  test('Snapshot empty list', () => {
    mockWebsocket();
    tree = mount(<Provider lobbyList={lobbyList}><LobbyList handleSetLobby={fn} /></Provider>)
    expect(tree.debug()).toMatchSnapshot();
  });

  test('Snapshot load on startup item', () => {
    const { lobbyList: lobbyListWebsocket, websocketFns } = mockWebsocket();

    let tree = mount(<Provider lobbyList={lobbyList}><LobbyList handleSetLobby={fn} /></Provider>);
    expect(lobbyListWebsocket).toBeCalled();
    websocketFns.lobbyListFn({
      'LName': {
        'lobbyName': 'S',
        'x': 10,
        'y': 10,
        'ships4n': 1,
        'ships3n': 1,
        'ships2n': 1,
        'ships1n': 1,
      }
    });
    tree.update();
    expect(tree.debug()).toMatchSnapshot();
  });

  test('Join to lobby', () => {
    const { lobbyList: lobbyListWebsocket, connectLobby, websocketFns } = mockWebsocket();
    
    let tree = mount(<Provider lobbyList={lobbyList}><LobbyList handleSetLobby={fn} /></Provider>);
    expect(lobbyListWebsocket).toBeCalled();
    websocketFns.lobbyListFn({
      'LName': {
        'lobbyName': 'S',
        'x': 10,
        'y': 10,
        'ships4n': 1,
        'ships3n': 1,
        'ships2n': 1,
        'ships1n': 1,
      }
    });

    const lobby = lobbyList.lobbys.get('LName')?.value as ILobbyStore;
    tree.update();
    tree.find('.btn-secondary').simulate('click');

    expect(connectLobby).toHaveBeenCalled();
    expect(connectLobby.mock.calls[0][0]).toBe('LName');
    expect(fn).toHaveBeenCalled();
    expect(fn.mock.calls[0][0]).toBe(GameStatus.SET_SHIPS);
    expect(fn.mock.calls[0][1]).toBe(lobby);
  });
});