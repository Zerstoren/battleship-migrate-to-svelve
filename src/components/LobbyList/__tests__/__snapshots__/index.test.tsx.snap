// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LobbyList Snapshot empty list 1`] = `
"<MobXProvider lobbyList={{...}}>
  <inject-with-lobbyList(Object) handleSetLobby={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
    <inject-with-lobbyList(Object) handleSetLobby={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} lobbyList={{...}}>
      <styled.h1>
        <h1 className=\\"sc-bdfBwQ display-3\\">
          Lobby list
        </h1>
      </styled.h1>
      <styled.table>
        <table className=\\"sc-gsTCUz table table-striped table-hover\\">
          <thead>
            <tr>
              <td />
              <td>
                Lobby name
              </td>
              <td>
                Field size
              </td>
              <td>
                Ships 4x/3x/2x/1x
              </td>
            </tr>
          </thead>
          <tbody />
        </table>
      </styled.table>
    </inject-with-lobbyList(Object)>
  </inject-with-lobbyList(Object)>
</MobXProvider>"
`;

exports[`LobbyList Snapshot load on startup item 1`] = `
"<MobXProvider lobbyList={{...}}>
  <inject-with-lobbyList(Object) handleSetLobby={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
    <inject-with-lobbyList(Object) handleSetLobby={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} lobbyList={{...}}>
      <styled.h1>
        <h1 className=\\"sc-bdfBwQ display-3\\">
          Lobby list
        </h1>
      </styled.h1>
      <styled.table>
        <table className=\\"sc-gsTCUz table table-striped table-hover\\">
          <thead>
            <tr>
              <td />
              <td>
                Lobby name
              </td>
              <td>
                Field size
              </td>
              <td>
                Ships 4x/3x/2x/1x
              </td>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <button className=\\"btn btn-secondary\\" onClick={[Function: onClick]}>
                  Join
                </button>
              </td>
              <td>
                S
              </td>
              <td>
                10
                x
                10
              </td>
              <td>
                1
                /
                1
                /
                1
                /
                1
              </td>
            </tr>
          </tbody>
        </table>
      </styled.table>
    </inject-with-lobbyList(Object)>
  </inject-with-lobbyList(Object)>
</MobXProvider>"
`;
